packages/core/react-loosely-lazy/src/lazy/components/client.tsx

```diff
@@ -22,95 +14,102 @@ import { PlaceholderFallbackRender } from '../placeholders/render';
 import { PlaceholderFallbackHydrate } from '../placeholders/hydrate';
 import { preloadAsset } from '../preload';
 
+import type { Status } from './types';
+import { useSubscription } from './utils';
+
+const { requestIdleCallback = setTimeout } = window;
+
 export function createComponentClient<C extends ComponentType<any>>({
   defer,
   deferred,
   dataLazyId,
   moduleId,
-  ssr,
 }: {
   defer: number;
   deferred: Deferred<C>;
   dataLazyId: string;
   moduleId: string;
-  ssr: boolean;
 }) {
   const ResolvedLazy = lazy(() => deferred.promise);
 
   return (props: ComponentProps<C>) => {
-    const { setFallback } = useContext(LazySuspenseContext);
-    const started = useRef(false);
-    const [, setState] = useState();
-    const until = useUntil();
+    // use a single piece of state to hold info about progress or eventually
+    // throw an error. We do change it via direct mutation as re-renders
+    // break Suspense in React 18, making it loose hydration state
+    const [status, bubbleError] = useState<Status>(() => ({
+      noWait: undefined,
+      phase: defer === PHASE.AFTER_PAINT ? false : true,
+      preloaded: defer === PHASE.AFTER_PAINT ? false : true,
+      started: false,
+    }));
 
-    const load = useRef(() => {
-      if (started.current) {
+    const load = () => {
+      if (status.started || !status.phase || !status.noWait) {
         return;
       }
 
-      started.current = true;
+      status.started = true;
       deferred.start().catch((err: Error) => {
         // Throw the error within the component lifecycle
         // refer to https://github.com/facebook/react/issues/11409
-        setState(() => {
+        bubbleError(() => {
           throw createLoaderError(err);
         });
       });
+    };
+
+    // Subscribe to LazyWait context, triggering load when until is true
+    useSubscription({
+      comparator: v => v === 1,
+      context: WaitContext,
+      key: 'noWait',
+      load,
+      status,
     });
 
-    if (defer === PHASE.LAZY) {
-      useEffect(() => {
-        if (until) {
-          load.current();
-        }
-      }, [until]);
-    } else {
-      const isOwnPhase = usePhaseSubscription(defer);
+    if (defer === PHASE.AFTER_PAINT) {
+      // Subscribe to LazyPhase context, triggering load when own phase starts
+      useSubscription({
+        comparator: v => v >= defer,
+        context: LazyPhaseContext,
+        key: 'phase',
+        load,
+        status,
+      });
 
+      // Schedule preloading as will be needed soon
       useMemo(() => {
-        if (isOwnPhase && until) {
-          load.current();
-        }
-      }, [isOwnPhase, until]);
-
-      if (defer === PHASE.AFTER_PAINT) {
-        // Start preloading as will be needed soon
-        useEffect(() => {
-          if (!isOwnPhase) {
-            return preloadAsset({
+        if (!status.preloaded) {
+          status.preloaded = true;
+          requestIdleCallback(() => {
```

Comment:
Given `useEffect` is not called in "real" suspended trees, move to `requestIdleCallback` so we do not impact render / critical path